File: solver_test.cpp
#include "solver.h"

double frand(double min, double max)
{
  double f = (double)std::rand()/RAND_MAX;
  return min + f*(max-min);
}

int main(int argc, char** argv)
{
	bool verbose = true;

	int seed = std::time(0);
	std::srand(seed);
	printf("IK Solver test starting with random seed %d\n", seed);

	kdlSolver solver("/home/dkebude/catkin_ws/src/panda_sim/panda_sim_description/urdf/panda_arm_hand.urdf","panda_link0","panda_hand", 100000, 125, 1e-6);

	int nj = solver.getNumJoints();

	KDL::JntArray j_min(nj);
	KDL::JntArray j_max(nj);
	solver.getJointLimits(j_min, j_max);

	std::vector<double> q_init (nj);
	std::vector<double> q_targ (nj);
	std::vector<double> p_targ (7); //quaternion

	q_init.at(0) = 1.15636;//(j_min(0)+j_max(0))/2;
	q_init.at(1) = 1.16243;//(j_min(1)+j_max(1))/2;
	q_init.at(2) = -0.61218;//(j_min(2)+j_max(2))/2;
	q_init.at(3) = -1.85039;//(j_min(3)+j_max(3))/2;
	q_init.at(4) = -2.88407;//(j_min(4)+j_max(4))/2;
	q_init.at(5) = 2.57014;//(j_min(5)+j_max(5))/2;
	q_init.at(6) = -2.53572;//(j_min(6)+j_max(6))/2;
	
	p_targ.at(0) = 0.485;//frand(j_min(0), j_max(0));
	p_targ.at(1) = 0.237;//frand(j_min(1), j_max(1));
	p_targ.at(2) = -0.028;//frand(j_min(2), j_max(2));
	p_targ.at(3) = 0.87;//frand(j_min(3), j_max(3));
	p_targ.at(4) = 0.117;//frand(j_min(4), j_max(4));
	p_targ.at(5) = 0.473;//frand(j_min(5), j_max(5));
	p_targ.at(6) = -0.076;//frand(j_min(6), j_max(6));
	normalizeQuaternion(p_targ);
	
	printf("\n------------------------\nTrying with pose IK Solver:\n------------------------\n");

	printf("\nInitial joint positions:\n------------------------\n");
	printDoubleVec(q_init);

	printf("\nTarget frame:\n------------------------\n");
	printDoubleVec(p_targ);

	printf("\nJoints as calculated by IK:\n------------------------\n");
	solver.solvePoseIk(q_init, q_targ, p_targ, verbose);
	printDoubleVec(q_targ);

	printf("\nFrame as calculated after IK:\n------------------------\n");
	solver.solveFkDouble(q_targ, p_targ);
	printDoubleVec(p_targ);

	printf("\n------------------------\nEuclidean distance between initial and target angles: %.5f\n------------------------\n", e_dist(q_init, q_targ));

	p_targ.at(0) = 0.485;//frand(j_min(0), j_max(0));
	p_targ.at(1) = 0.237;//frand(j_min(1), j_max(1));
	p_targ.at(2) = -0.028;//frand(j_min(2), j_max(2));
	p_targ.at(3) = 0.87;//frand(j_min(3), j_max(3));
	p_targ.at(4) = 0.117;//frand(j_min(4), j_max(4));
	p_targ.at(5) = 0.473;//frand(j_min(5), j_max(5));
	p_targ.at(6) = -0.076;//frand(j_min(6), j_max(6));
	normalizeQuaternion(p_targ);

	printf("\n------------------------\nTrying with position only IK Solver:\n------------------------\n");

	printf("\nInitial joint positions:\n------------------------\n");
	printDoubleVec(q_init);

	printf("\nTarget frame:\n------------------------\n");
	printDoubleVec(p_targ);

	printf("\nJoints as calculated by IK:\n------------------------\n");
	solver.solvePosOnlyIk(q_init, q_targ, p_targ, verbose);
	printDoubleVec(q_targ);

	printf("\nFrame as calculated after IK:\n------------------------\n");
	solver.solveFkDouble(q_targ, p_targ);
	printDoubleVec(p_targ);

	printf("\n------------------------\nEuclidean distance between initial and target angles: %.5f\n------------------------\n", e_dist(q_init, q_targ));

	p_targ.at(0) = 0.485;//frand(j_min(0), j_max(0));
	p_targ.at(1) = 0.237;//frand(j_min(1), j_max(1));
	p_targ.at(2) = -0.028;//frand(j_min(2), j_max(2));
	p_targ.at(3) = 0.87;//frand(j_min(3), j_max(3));
	p_targ.at(4) = 0.117;//frand(j_min(4), j_max(4));
	p_targ.at(5) = 0.473;//frand(j_min(5), j_max(5));
	p_targ.at(6) = -0.076;//frand(j_min(6), j_max(6));
	normalizeQuaternion(p_targ);

	printf("\n------------------------\nTrying with hybrid IK Solver:\n------------------------\n");

	double jnt_dist_threshold = 1.5;
	verbose = false;

	printf("\nInitial joint positions:\n------------------------\n");
	printDoubleVec(q_init);

	printf("\nTarget frame:\n------------------------\n");
	printDoubleVec(p_targ);

	printf("\nJoints as calculated by IK:\n------------------------\n");
	solver.solveHybridIk(q_init, q_targ, p_targ, jnt_dist_threshold, verbose);
	printDoubleVec(q_targ);

	printf("\nFrame as calculated after IK:\n------------------------\n");
	solver.solveFkDouble(q_targ, p_targ);
	printDoubleVec(p_targ);

	printf("\n------------------------\nEuclidean distance between initial and target angles: %.5f\n------------------------\n", e_dist(q_init, q_targ));

	return 0;
}

File: PyIkSolver/solver_test.py
#!/usr/bin/env python
import rospy
from solver import *

def main():
	verbose = False

	solver = kdlSolver("/home/dkebude/catkin_ws/src/panda_sim/panda_sim_description/urdf/panda_arm_hand.urdf","panda_link0","panda_hand", 100000, 125, 0.000001)

	num_joints = solver.getNumJoints()

	q_targ = DoubleVector(num_joints)

	q_init = DoubleVector(num_joints)
	q_init[0] = 1.15636
	q_init[1] = 1.16243
	q_init[2] = -0.61218
	q_init[3] = -1.85039
	q_init[4] = -2.88407
	q_init[5] = 2.57014
	q_init[6] = -2.53572

	p_targ = DoubleVector(7)
	p_targ[0] = 0.485
	p_targ[1] = 0.237
	p_targ[2] = -0.028
	p_targ[3] = 0.87
	p_targ[4] = 0.117
	p_targ[5] = 0.473
	p_targ[6] = -0.076
	normalizeQuaternion(p_targ)

	print "\n------------------------\nTrying with pose IK Solver:\n------------------------\n"
	print "\nInitial joint positions:\n------------------------\n"
	printDoubleVec(q_init)

	print "\nTarget frame:\n------------------------\n"
	printDoubleVec(p_targ)

	print "\nJoints as calculated by IK:\n------------------------\n"
	solver.solvePoseIk(q_init, q_targ, p_targ, verbose)
	printDoubleVec(q_targ)

	print "\nFrame as calculated after IK:\n------------------------\n"
	solver.solveFkDouble(q_targ, p_targ)
	printDoubleVec(p_targ)

	print "\n------------------------\nEuclidean distance between initial and target angles: %.5f\n------------------------\n"%e_dist(q_init, q_targ)

	p_targ[0] = 0.485
	p_targ[1] = 0.237
	p_targ[2] = -0.028
	p_targ[3] = 0.87
	p_targ[4] = 0.117
	p_targ[5] = 0.473
	p_targ[6] = -0.076
	normalizeQuaternion(p_targ)

	print "\n------------------------\nTrying with position only IK Solver:\n------------------------\n"
	print "\nInitial joint positions:\n------------------------\n"
	printDoubleVec(q_init)

	print "\nTarget frame:\n------------------------\n"
	printDoubleVec(p_targ)

	print "\nJoints as calculated by IK:\n------------------------\n"
	solver.solvePosOnlyIk(q_init, q_targ, p_targ, verbose)
	printDoubleVec(q_targ)

	print "\nFrame as calculated after IK:\n------------------------\n"
	solver.solveFkDouble(q_targ, p_targ)
	printDoubleVec(p_targ)

	print "\n------------------------\nEuclidean distance between initial and target angles: %.5f\n------------------------\n"%e_dist(q_init, q_targ)

	p_targ[0] = 0.485
	p_targ[1] = 0.237
	p_targ[2] = -0.028
	p_targ[3] = 0.87
	p_targ[4] = 0.117
	p_targ[5] = 0.473
	p_targ[6] = -0.076
	normalizeQuaternion(p_targ)

	jnt_dist_th = 1.5
	verbose = False

	print "\n------------------------\nTrying with hybrid IK Solver:\n------------------------\n"
	print "\nInitial joint positions:\n------------------------\n"
	printDoubleVec(q_init)

	print "\nTarget frame:\n------------------------\n"
	printDoubleVec(p_targ)

	print "\nJoints as calculated by IK:\n------------------------\n"
	solver.solveHybridIk(q_init, q_targ, p_targ, jnt_dist_th, verbose)
	printDoubleVec(q_targ)

	print "\nFrame as calculated after IK:\n------------------------\n"
	solver.solveFkDouble(q_targ, p_targ)
	printDoubleVec(p_targ)

	print "\n------------------------\nEuclidean distance between initial and target angles: %.5f\n------------------------\n"%e_dist(q_init, q_targ)

if __name__ == '__main__':
	try:
	    main()
	except rospy.ROSInterruptException:
	    pass